File: Tests/Categories/TestUIViewController.m
@import UIKit;
@import XCTest;
#import <PromiseKit/PromiseKit.h>
#import "UIViewController+AnyPromise.h"


@interface MyViewController: UIViewController
@property AnyPromise *promise;
@end
@implementation MyViewController
@end


@implementation Test_UIViewController_ObjC: XCTestCase {
    UIViewController *rootvc;
}

- (void)setUp {
    rootvc = [UIApplication sharedApplication].keyWindow.rootViewController = [UIViewController new];
}

- (void)tearDown {
    [UIApplication sharedApplication].keyWindow.rootViewController = nil;
}

// view controller is presented and dismissed when promise is resolved
- (void)test1 {
    id ex = [self expectationWithDescription:@""];

    PMKResolver resolve;

    MyViewController *myvc = [MyViewController new];
    myvc.promise = [[AnyPromise alloc] initWithResolver:&resolve];
    [rootvc promiseViewController:myvc animated:NO completion:nil].then(^{
        // seems to take another tick for the dismissal to complete
    }).then(^{
        [ex fulfill];
    });

    XCTAssertNotNil(rootvc.presentedViewController);

    PMKAfter(1).then(^{
        resolve(@1);
    });

    [self waitForExpectationsWithTimeout:10 handler:nil];

    XCTAssertNil(rootvc.presentedViewController);
}

// view controller is not presented if promise is resolved
- (void)test2 {
    MyViewController *myvc = [MyViewController new];
    myvc.promise = [AnyPromise promiseWithValue:nil];
    [rootvc promiseViewController:myvc animated:NO completion:nil];

    XCTAssertNil(rootvc.presentedViewController);
}

// promise property must be promise
- (void)test3 {
    id ex = [self expectationWithDescription:@""];

    MyViewController *myvc = [MyViewController new];
    myvc.promise = (id) @1;
    [rootvc promiseViewController:myvc animated:NO completion:nil].catch(^(id err){
        [ex fulfill];
    });

    XCTAssertNil(rootvc.presentedViewController);

    [self waitForExpectationsWithTimeout:10 handler:nil];

    XCTAssertNil(rootvc.presentedViewController);
}

// promise property must not be nil
- (void)test4 {
    id ex = [self expectationWithDescription:@""];

    MyViewController *myvc = [MyViewController new];
    [rootvc promiseViewController:myvc animated:NO completion:nil].catch(^(id err){
        [ex fulfill];
    });

    XCTAssertNil(rootvc.presentedViewController);

    [self waitForExpectationsWithTimeout:10 handler:nil];

    XCTAssertNil(rootvc.presentedViewController);
}

// view controller must have a promise property
- (void)test5 {
    id ex = [self expectationWithDescription:@""];

    UIViewController *vc = [UIViewController new];
    [rootvc promiseViewController:vc animated:NO completion:nil].catch(^(id err){
        [ex fulfill];
    });

    XCTAssertNil(rootvc.presentedViewController);

    [self waitForExpectationsWithTimeout:10 handler:nil];

    XCTAssertNil(rootvc.presentedViewController);
}

// promised nav controllers use their root vc’s promise property
- (void)test6 {
    id ex = [self expectationWithDescription:@""];

    PMKResolver resolve;

    MyViewController *myvc = [MyViewController new];
    UINavigationController *nc = [[UINavigationController alloc] initWithRootViewController:myvc];
    myvc.promise = [[AnyPromise alloc] initWithResolver:&resolve];
    [rootvc promiseViewController:nc animated:NO completion:nil].then(^(id obj){
        XCTAssertEqualObjects(@1, obj);
    }).then(^{
        [ex fulfill];
    });

    XCTAssertNotNil(rootvc.presentedViewController);

    PMKAfter(1).then(^{
        resolve(@1);
    });

    [self waitForExpectationsWithTimeout:10 handler:nil];

    XCTAssertNil(rootvc.presentedViewController);
}

@end

File: Tests/Categories/TestUIViewController.swift
import PromiseKit
import UIKit
import XCTest

private let dummy = 1_234_765


class Test_UIViewController_Swift: UIKitTestCase {

    func test_rejects_if_vc_has_no_promise_property() {
        let ex = expectation(description: "")
        let p: Promise<Int> = rootvc.promiseViewController(UIViewController(), animated: false)
        p.catch { error in
            if case UIViewController.Error.notPromisable = error {
                ex.fulfill()
            }
        }
        waitForExpectations(timeout: 1, handler: nil)
    }

    func test_rejects_if_promise_property_returns_nil() {
        let ex = expectation(description: "")
        let p: Promise<Int> = rootvc.promiseViewController(MockViewController(), animated: false)
        p.catch { error in
            if case UIViewController.Error.nilPromisable = error {
                ex.fulfill()
            }
        }
        waitForExpectations(timeout: 1, handler: nil)
    }

    func test_rejects_if_promise_property_casts_wrongly() {
        let ex = expectation(description: "")
        let my = MockViewController()
        my.promise = Promise.resolved(value: true)
        let p: Promise<Int> = rootvc.promiseViewController(my, animated: false)
        p.catch { err in
            if case UIViewController.Error.notGenericallyPromisable = err {
                ex.fulfill()
            }
        }
        waitForExpectations(timeout: 1, handler: nil)
    }

    func test_resolved_promise_property_means_vc_does_not_appear() {
        let ex = expectation(description: "")
        let my = MockViewController()
        my.promise = Promise.resolved(value: dummy)
        rootvc.promiseViewController(my, animated: false).then { (x: Int) -> Void in
            XCTAssertFalse(my.appeared)
            XCTAssertEqual(x, dummy)
            ex.fulfill()
        }
        waitForExpectations(timeout: 1, handler: nil)
    }

    func test_vc_dismisses_once_promise_is_resolved() {
        let ex = expectation(description: "")
        let my = MockViewController()
        let (promise, resolve, _) = Promise<Int>.pending()
        my.promise = promise
        rootvc.promiseViewController(my, animated: false).then { (x: Int) -> Void in
            XCTAssertTrue(my.appeared)
            XCTAssertEqual(x, dummy)
            ex.fulfill()
        }
        after(interval: 0).then {
            resolve(dummy)
        }
        waitForExpectations(timeout: 1, handler: nil)
    }

    func test_nav_controllers_use_their_root_vc_promise_property() {
        let ex = expectation(description: "")
        let nc = UINavigationController()
        let my = MockViewController()
        my.promise = after(interval: 0.1).then{ dummy }
        nc.viewControllers = [my]
        rootvc.promiseViewController(nc, animated: false).then { (x: Int) -> Void in
            XCTAssertTrue(my.appeared)
            XCTAssertEqual(x, dummy)
            ex.fulfill()
        }
        waitForExpectations(timeout: 1, handler: nil)
    }
}



private class MockViewController: UIViewController, Promisable {
    @objc var promise: AnyObject! = nil

    var appeared = false

    private override func viewDidAppear(_ animated: Bool) {
        appeared = true
    }
}
